[AMDGPU] Run LowerLDS at the end of the fullLTO pipeline

Pierre-vh opened the pull request:
This change allows us to use `--lto-partitions` in some cases (not at all guaranteed it works perfectly), as LDS is lowered before the module is split for parallel codegen.

LowerrLDS doesn't support being ran twice because it'll think the lowered LDS GVs are "absolute addresses" LDS which aren't supported, so I just added a module flag to detect multiple runs.

We must run LowerLDS before splitting modules as it needs to see all callers of functions with LDS to properly lower them.

Pierre-vh commented:
Just a thought: if we're not okay with changing the order of LowerModuleLDS for gpu-rdc (because this affects all full LTO builds), I can try adding a parameter to the callback to tell us if we're doing module splitting or not, and if we're not, we just run things as usual.

JonChesterfield commented:
Moving the pass is probably fine - it should be kept adjacent to promotealloca but otherwise doesn't matter much.

The flag is ugly but it'll take some effort to make the pass properly composable, in the sense that it can be run repeatedly or on separate TUs before on the whole module.

Similarly hacky to the flag, we could look for the magic lookup table it builds (has a known name) and skip the pass if it already exists.

arsenm commented:
> Moving the pass is probably fine - it should be kept adjacent to promotealloca but otherwise doesn't matter much. Reasonable chance promotealloca will need fixes too.


BTW what happened to the patches to allow module LDS to report to promote-alloca the available LDS budget?


> 
> The flag is ugly but it'll take some effort to make the pass properly composable, in the sense that it can be run repeatedly or on separate TUs before on the whole module.

That would be nice

Pierre-vh commented:
Is this okay to land?

Pierre-vh commented:
ping

arsenm reviewed (commented):
How difficult would it be to support absolute addresses in the pass?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULowerModuleLDSPass.cpp:
```diff
@@ -1088,6 +1096,15 @@ class AMDGPULowerModuleLDS {
+    // This pass may run twice in a full LTO pipeline.
+    //
+    // If we ran it early, we'll have added metadata to skip next runs.
+    if (M.getModuleFlag("amdgcn.lowered_module_lds"))
```
Rather than communicating this with the flag, better to just not add it in the pipeline configuration?

Pierre-vh replied:
I think there's different TargetMachine instances in this case, we have no way to tell if we're running the backend as part of a full lto partitioned module (= LowerLDS ran) or if we're running it as usual (= we'd like to keep LowerLDS in the usual place). The module flag is the only way to reliably communicate that I can think of

Pierre-vh commented:
> How difficult would it be to support absolute addresses in the pass?

I'm not sure, that's a question for @JonChesterfield I believe :)
In this case I'm just interested in making the pass able to run more than once to enable some experiments with ``--lto-partitions`

Pierre-vh commented:
ping

Pierre-vh commented:
ping - can it land?

Pierre-vh commented:
Ping

JonChesterfield commented:
> BTW what happened to the patches to allow module LDS to report to promote-alloca the available LDS budget?

I think the machinery landed and the promote alloca patch got lost somewhere around the phab -> github transition. There are some limitations around there that could be stamped out, e.g. there's no particular reason the alloca->lds translation has to be restricted to kernels. I'll push it back onto the work list.

> How difficult would it be to support absolute addresses in the pass?

The LDS lowering pass (currently) crawls the IR module, writes "allocates N bytes" on kernels and absolute address metadata on variables. It's partly set up like that in order to support application developers writing absolute address constraints on globals as that seems generally useful and someone thought it was a plausible feature request in one of the meetings.

As far as allocating goes, making it composable means treating the kernel as a bump allocator. That's definitely fine.

There are some sketchy semantics and codgen implications that falls out.

I think the conclusion of the design choices will be that a variable with an absolute address cannot be moved, kernels will not try to allocate extra memory to encompass it, data layout will be generally compromised relative to the single application case.

Given two modules which have independently had LDS lowering applied, llvm-link of the two is generally going to be unsound, for the same reasons that machine code linking of independently lowered LDS is not viable.

So while "handle absolute addresses" is relatively straightforward from the perspective of the LDS lowering pass, it's going to be interpreted as meaning LDS lowering can be arbitrarily composed as opposed to the very limited situations in which it's actually valid. In particular, a cleverer LDS pass will make the current error message go away, and do something entirely locally correct, and the emergent behaviour will be broken nonsense as soon as someone steps slightly away from the thinlto workflow, probably without any diagnostics.

I would propose the following:

1. Check for any LDS variables with absolute addresses.
2. If all LDS variables have absolute addresses, no work to do, return. Success.
3. If none have absolute addresses, run the pass. Add a postcondition that ever LDS variable has an absolute address if it's not already there.
4. If some have absolute addresses, and some do not, fatal_error with some message along the lines of "no deal, this stuff doesn't compose properly".

Other interesting ideas are to refuse to compile code containing external LDS variables, which is currently left to the backend to report iirc.

The right thing is to change the LDS lowering scheme such that it does work in the presence of machine code linking. That is difficult to do and probably involves patching the linker and/or the loaders. However that will then work however code is spliced together, modulo kernels refusing to launch because there's only 64k or so of LDS available and machine code linking is going to compromise data layout efficiency badly.

The above suggestion is better than trying to pass state around to disable subsequent calls to the pass. If you want to do the magic implicit disabling instead, do it in the pass pipeline control, not in the pass itself.

The thinlto workflow will hit case 3. on each piece, then hit case 2. if I'm understanding correctly. Hopefully it doesn't duplicate global variables while splitting the module as that definitely won't work.

Pierre-vh committed:
[AMDGPU] Run LowerLDS at the end of the fullLTO pipeline

This change allows us to use `--lto-partitions` in some cases (not guaranteed it works perfectly), as LDS is lowered before the module is split for parallel codegen.

Pierre-vh commented:
This now just adds the pass to the FullLTO pipeline
@arsenm

arsenm commented:
Can we get some tests for this? In particular I would like to be sure that we still compile correctly -O0 + LTO. Do we have tests that run the module lowering multiple times?

Pierre-vh commented:
> Can we get some tests for this? In particular I would like to be sure that we still compile correctly -O0 + LTO. Do we have tests that run the module lowering multiple times?

 #81729 added one

Pierre-vh committed:
add lto pipeline test

Pierre-vh commented:
This is causing `LLVM ERROR: Module cannot mix absolute and non-absolute LDS GVs` on the openmp buildbot, and I have no idea why: https://lab.llvm.org/buildbot/#/builders/193/builds/48532/steps/6/logs/FAIL__libomptarget____amdgcn-amd-amdhsa__cuda_no_d
I cannot reproduce it locally either

JonChesterfield commented:
Running the pass on the end of full lto doesn't work. There's lots of ways to break that, it was specifically thin lto that carefully carved out an exception. This has just shown up in fortran, @ergawy pinged me about it.

I'll raise an issue to get better visibility on it.
